### PR TITLE
Multi change Pull request

### PR DESCRIPTION
**Pull request by source83**
Hi, this request includes several changes.
1. Starters for user authentication. (Waiting on James for the front end for it, but it works)
2. Access logging by date and ip for statistical purposes.
3. Updating of witch users are connected and when they originally started the connection. (Starters for a friends list and more.)

There is more to be done and more cleanup of my code, this is just a initial pull request for now.

Regards,
Source

### CODE REVIEW

**Comment by sirkitree** [comment]
There are some merge conflict that you should fix, but this is looking very interesting. 

I'll give a more in depth review, but note first that it's much easier to get pull requests into a project where the maintainer has limited time if you split out things into their own discrete pull requests per feature rather than putting 3 features into 1. :wink:

File: api.txt [hunk 1]
```diff
@@ -79,3 +79,26 @@ The followed up with a move user_moved event
...
 {"method":"user_portal", "data":{"userId":"LL","roomId":"345678354764987457","url":"http://...", "pos":[0,0,0], "fwd":[0,1,0]}}
+
+
+Request how many users are connected to the server:
+{"method":"getusersonline"}
+
+Returns:
+{"data": 2, "method": "usersonline"}
+
+To get user count in the current room (NOT WORKING YET, WILL RETURN 0 FOR THE TIME BEING)
```
Comment (sirkitree) [inline api.txt]: Personally, I typically prefer to not commit non-working code to a project, but I'll leave that up to @lisa-lionheart

File: config.js [hunk 2]
```diff
@@ -15,6 +15,25 @@ module.exports = {
...
+	/* Update user authentication information interval in minutes */
+	UserInfo_updateInterval: 5,
+	/*
+		server mode 1: Free for all, anybody can log on with any userId
+		server mode 2: Registered userId's will trigger a password request
+		server mode 3: Only registered userId's are allowed to connect
+	*/
+	server_mode: 2,
+
+	/* wether or not to log connection stats */
```
Comment (sirkitree) [inline config.js]: sp: whether not wether

File: server.js [hunk 3]
```diff
@@ -118,6 +138,70 @@ Server.prototype.onConnect = function(socket) {
...
+		host     : config.MySQL_Hostname,
+		user     : config.MySQL_Username,
+		password : config.MySQL_Password,
+	});
+
+	dbcon.connect(function(err) {
+		// connected! (unless `err` is set)
+	});
+	var post = {ip: addr};
+	dbcon.query('INSERT INTO access_statistics SET ?', post,  function(err, results) {
```
Comment (sirkitree) [inline server.js]: perhaps this should go inside of the `connect()` when `err` is not set?

File: server.js [hunk 4]
```diff
@@ -118,6 +138,70 @@ Server.prototype.onConnect = function(socket) {
...
+};
+
+Server.prototype.updateUserList = function() {
+	log.info("Updating userlist");
+	var dbcon = mysql.createConnection({
+		database : config.MySQL_Database,
+		host     : config.MySQL_Hostname,
+		user     : config.MySQL_Username,
+		password : config.MySQL_Password,
+	});
```
Comment (sirkitree) [inline server.js]: since this is used more than once it might be a good candidate for a new method

File: server.js [hunk 5]
```diff
@@ -118,6 +138,70 @@ Server.prototype.onConnect = function(socket) {
...
+
+	while(userList.length > 0) {
+		userList.pop();
+	}
+
+	dbcon.connect(function(err) {
+		// connected! (unless `err` is set)
+	});
+
+	dbcon.query('SELECT * FROM usernames ORDER BY user DESC', function(err, rows, fields) {
```
Comment (sirkitree) [inline server.js]: again, should this go inside of `connect()`?

File: src/Session.js [hunk 6]
```diff
@@ -7,9 +9,13 @@ function Session(server, socket) {
     var self = this;
 
     this._socket = socket;
+
+	this._userLocked = false;
```
Comment (sirkitree) [inline src/Session.js]: looks like your spacing / tabs might be different than the original project. I'd suggest conforming to what the project is already using.

File: src/Session.js [hunk 7]
```diff
@@ -42,9 +68,22 @@ Session.prototype.clientError = function(message) {
...
+						'enter_room', 
+						'move', 
+						'chat', 
+						'portal', 
+						'usersonline', 
+						'getusersonline', 
+						'usersinroom', 
+						'getusersinroom',
+						'passwordrequest'
+					];
```
Comment (sirkitree) [inline src/Session.js]: holy indentation Batman!

Comment (source83) [inline src/Session.js]: I don't really care about the esthetics as long as it runs.

File: server.js (hunk 5 above)
Comment (source83) [inline server.js]: what connect() ?

File: server.js (hunk 4 above)
Comment (source83) [inline server.js]: as in a json method ?

**Pull request closed without merging**